searx/version.py

```diff
@@ -1,26 +1,150 @@
+# ( MIT License )
+
+import os
+import shlex
+import subprocess
+
+
+class ShellException(Exception):
+
+    def __init__(self, command, exit_code, stdout, stderr):
+        super().__init__(self, stderr)
+        self.command = command
+        self.exit_code = exit_code
+        self.stdout = stdout
+        self.stderr = stderr
+
+
+def no_git_env(env):
+    # adapted from pre-commit
+    # Too many bugs dealing with environment variables and GIT:
+    # https://github.com/pre-commit/pre-commit/issues/300
+    # In git 2.6.3 (maybe others), git exports GIT_WORK_TREE while running
+    # pre-commit hooks
+    # In git 1.9.1 (maybe others), git exports GIT_DIR and GIT_INDEX_FILE
+    # while running pre-commit hooks in submodules.
+    # GIT_DIR: Causes git clone to clone wrong thing
+    # GIT_INDEX_FILE: Causes 'error invalid object ...' during commit
+    return {
+        k: v
+        for k, v in env.items()
+        if not k.startswith("GIT_")
+        or k in ("GIT_EXEC_PATH", "GIT_SSH", "GIT_SSH_COMMAND")
+    }
+
+
+def _popen_pipes(cmd, cwd):
+    # pylint: disable=consider-using-with
+    return subprocess.Popen(
+        cmd,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        cwd=str(cwd),
+        env=dict(
+                no_git_env(os.environ),
+                # os.environ,
+                # try to disable i18n
+                LC_ALL="C",
+                LANGUAGE="",
+        ),
+    )
+
+
+def ensure_stripped_str(str_or_bytes):
+    if isinstance(str_or_bytes, str):
+        return str_or_bytes.strip()
+    return str_or_bytes.decode("utf-8", "surrogateescape").strip()
+
+
+def do_ex(cmd, cwd="."):
+    if os.name == "posix" and not isinstance(cmd, (list, tuple)):
+        cmd = shlex.split(cmd)
+
+    p = _popen_pipes(cmd, cwd)
+    out, err = p.communicate()
+    return ensure_stripped_str(out), ensure_stripped_str(err), p.returncode
+
+
+def do_ex_safe(cmd, cwd="."):
+    out, err, exit_code = do_ex(cmd, cwd)
+    if exit_code != 0:
+        raise ShellException(cmd, exit_code, out, err)
+    return out
+
+
+def get_public_git_url(git_url):
+    if git_url.startswith('git@'):
+        git_url = git_url.replace(':', '/', 2).replace('git@', 'https://', 1)
+    if git_url.endswith('.git'):
+        return git_url.replace('.git', '', 1)
+    return git_url
+
+
+def get_git_url_and_branch():
+    try:
+        ref = do_ex_safe('git rev-parse --abbrev-ref @{upstream}')
+    except ShellException:
+        ref = do_ex_safe('git rev-parse --abbrev-ref master@{upstream}')
+    origin, git_branch = ref.split('/', 1)
+    git_url = do_ex_safe(f'git remote get-url {origin}')
+    return get_public_git_url(git_url), git_branch
+
+
+def get_git_version():
+    git_version = do_ex_safe(r'git describe --match "v[0-9]*\.[0-9]*\.[0-9]*" HEAD')
+    git_version_s = git_version.split('-')
+    git_version_s[0] = git_version_s[0][1:]
+    if len(git_version_s) == 3:
+        git_version_s[2] = git_version_s[2][1:]
```

Comment:
Juggling with the index in the list is hard to read I recommend to name the values:

```python
tag = proc.stdout.decode().strip()
tag_version, tag_distance, tag_commit = tag.split('-')
```

Reply:
Done:
https://github.com/searxng/searxng/blob/e5d1cd3c9deccfd2099e48baed855dcc00113ed0/searx/version.py#L93-L106

Note: If the last commit is a tag `git describe` will return only the tag without distance or commit hash.